File: src/topcoder/PrimeAnagrams.java
package topcoder;

import java.util.ArrayList;
import java.util.Arrays;

public class PrimeAnagrams {
	public static void main(String[] args) {
		int[] primes = new PrimeAnagrams().primes("39151687");
		System.out.println(primes[0]);
		System.out.println(primes[1]);
		System.out.println(primes[2]);
	}

	private boolean[] primesArr = new boolean[1000010];
	private ArrayList<Integer> primes = new ArrayList<>();
	private int[] result = new int[3];
	private int min = Integer.MAX_VALUE;

	public int[] primes(String anagram) {
		calculatePrimes();
		permute(anagram.toCharArray(), 0, anagram.length());
		Arrays.sort(result);
		return min < Integer.MAX_VALUE ? result : new int[0];
	}

	private void permute(char[] anagram, int l, int r) {
		if (l >= r) {
			String s = String.valueOf(anagram);
			for (int i = 1; i < r; i++) {
				for (int j = i + 1; j < r; j++) {
					findPrimes(s.substring(0, i), s.substring(i, j), s.substring(j));
				}
			}
		} else {
			for (int i = l; i < r; i++) {
				swap(anagram, l, i);
				permute(anagram, l + 1, r);
				swap(anagram, l, i);
			}
		}
	}

	private void swap(char[] anagram, int l, int i) {
		char temp = anagram[l];
		anagram[l] = anagram[i];
		anagram[i] = temp;
	}

	private void findPrimes(String x, String y, String z) {
		if (x.charAt(0) == '0') return;
		if (y.charAt(0) == '0') return;
		if (z.charAt(0) == '0') return;

		int X = Integer.parseInt(x);
		int Y = Integer.parseInt(y);
		int Z = Integer.parseInt(z);

		if (primesArr[X] && primesArr[Y] && primesArr[Z] && X * Y * Z < min) {
			min = X * Y * Z;
			result[0] = X;
			result[1] = Y;
			result[2] = Z;
		}
	}

	private void calculatePrimes() {
		for (int i = 2; i < 1000000; i++) {
			if (isPrime(i)) {
				primes.add(i);
				primesArr[i] = true;
			}
		}
	}

	private boolean isPrime(int n) {
		for (int i = 0; i < primes.size() && primes.get(i) * primes.get(i) <= n; i++) {
			if (n % primes.get(i) == 0) {
				return false;
			}
		}
		return true;
	}
}

File: cpp/topcoder/PrimeAnagrams.cc
#include <algorithm>
#include <cmath>
#include <iostream>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;

class PrimeAnagrams {
   private:
    static constexpr int INF = 1e6 + 10;
    int min = INT32_MAX;
    vector<int> result;
    vector<int> prime_vector;

    void CalculatePrimes(int max) {
        prime_vector.push_back(2);
        for (int i = 3; i <= max; i += 2) {
            if (IsPrime(i)) {
                prime_vector.push_back(i);
            }
        }
    }

    bool IsPrime(int n) {
        if (n <= 1) return false;
        int max = sqrt(n);
        for (int i = 0; i < prime_vector.size() && prime_vector[i] <= max;
             i++) {
            if (n % prime_vector[i] == 0) {
                return false;
            }
        }
        return true;
    }

    void ValidatePrimes(string first, string second, string third) {
        if (first[0] == '0' || second[0] == '0' || third[0] == '0') return;
        int x = stoi(first);
        int y = stoi(second);
        int z = stoi(third);
        if (IsPrime(x) && IsPrime(y) && IsPrime(z) && (x * y * z < min)) {
            min = x * y * z;
            result[0] = x;
            result[1] = y;
            result[2] = z;
        }
    }

    void PermuteAndCalculate(string anagram, int start, int end) {
        if (start == end) {
            for (int i = 1; i < end; i++) {
                for (int j = i + 1; j < end; j++) {
                    ValidatePrimes(anagram.substr(0, i),
                                   anagram.substr(i, j - i),
                                   anagram.substr(j, end));
                }
            }
        } else {
            for (int i = start; i < end; i++) {
                swap(anagram, start, i);
                PermuteAndCalculate(anagram, start + 1, end);
                swap(anagram, start, i);
            }
        }
    }

    void swap(string &s, int i, int j) {
        char temp = s[i];
        s[i] = s[j];
        s[j] = temp;
    }

   public:
    PrimeAnagrams() : result(3) { CalculatePrimes(INF); }

    vector<int> primes(string anagram) {
        PermuteAndCalculate(anagram, 0, anagram.length());
        sort(result.begin(), result.end());
        return (min != INT32_MAX) ? result : vector<int>();
    }
};

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    string s;
    cin >> s;
    PrimeAnagrams p;
    vector<int> result = p.primes(s);
    cout << result[0] << " " << result[1] << " " << result[2] << endl;

    return 0;
}
